{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c56ca3c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix\n",
    "from sklearn.model_selection import GridSearchCV\n",
    "from pyod.models.ocsvm import OCSVM\n",
    "from pyod.models.pca import PCA\n",
    "from pyod.models.knn import KNN\n",
    "from pyod.models.iforest import IForest\n",
    "from pyod.utils.data import evaluate_print\n",
    "import time\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b13f76b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load and preprocess the dataset for training\n",
    "df_train = pd.read_csv('train1.csv')  # Replace 'your_train_dataset.csv' with the actual file name\n",
    "train_data = df_train.drop(['Time', 'BLANK1', 'BLANK2', 'BLANK3'], axis=1)\n",
    "train_data.dropna(axis=0, inplace=True)\n",
    "X_train = train_data.values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "605633f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_search.py:925: UserWarning: One or more of the test scores are non-finite: [nan nan nan nan nan nan nan nan nan nan nan nan]\n",
      "  category=UserWarning\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_search.py:925: UserWarning: One or more of the test scores are non-finite: [nan nan nan]\n",
      "  category=UserWarning\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_search.py:925: UserWarning: One or more of the test scores are non-finite: [nan nan nan nan]\n",
      "  category=UserWarning\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:700: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 685, in _score\n",
      "    scores = scorer(estimator, X_test)\n",
      "TypeError: __call__() missing 1 required positional argument: 'y_true'\n",
      "\n",
      "  UserWarning,\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\sklearn\\model_selection\\_search.py:925: UserWarning: One or more of the test scores are non-finite: [nan nan nan]\n",
      "  category=UserWarning\n"
     ]
    }
   ],
   "source": [
    "# Define the range of hyperparameters for each model\n",
    "#ocsvm_params = {'nu': [0.1, 0.3, 0.5]}\n",
    "ocsvm_params = {'kernel': ['linear', 'rbf'], 'gamma': ['scale', 'auto'], 'nu': [0.1, 0.2, 0.3]}\n",
    "pca_params = {'n_components': [2, 4, 6]}\n",
    "knn_params = {'n_neighbors': [3, 5, 7, 9 ]}\n",
    "iforest_params = {'n_estimators': [50, 100, 150]}\n",
    "\n",
    "# Perform hyperparameter tuning using grid search\n",
    "models = {'OCSVM': (OCSVM(), ocsvm_params),\n",
    "          'PCA': (PCA(), pca_params),\n",
    "          'kNN': (KNN(), knn_params),\n",
    "          'Isolation Forest': (IForest(), iforest_params)}\n",
    "\n",
    "best_models = {}\n",
    "\n",
    "for model_name, (model, param_grid) in models.items():\n",
    "    grid_search = GridSearchCV(model, param_grid, scoring='accuracy', cv=5)\n",
    "    grid_search.fit(X_train)\n",
    "    best_model = grid_search.best_estimator_\n",
    "    best_models[model_name] = best_model\n",
    "\n",
    "# Assign the best models to variables\n",
    "best_ocsvm = best_models['OCSVM']\n",
    "best_pca = best_models['PCA']\n",
    "best_knn = best_models['kNN']\n",
    "best_iforest = best_models['Isolation Forest']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "26cf6bc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load and preprocess the test datasets\n",
    "df_test_01 = pd.read_csv('test1.csv')  # Replace 'your_test_dataset_01.csv' with the actual file name\n",
    "df_test_02 = pd.read_csv('test2.csv')  # Replace 'your_test_dataset_02.csv' with the actual file name\n",
    "test_data_01 = df_test_01.drop(['Time', 'BLANK1', 'BLANK2', 'BLANK3'], axis=1)\n",
    "test_data_02 = df_test_02.drop(['Time', 'BLANK1', 'BLANK2', 'BLANK3'], axis=1)\n",
    "test_data_01.dropna(axis=0, inplace=True)\n",
    "test_data_02.dropna(axis=0, inplace=True)\n",
    "X_test_01 = test_data_01.values\n",
    "X_test_02 = test_data_02.values\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d0b4834b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\ipykernel_launcher.py:18: RuntimeWarning: invalid value encountered in longlong_scalars\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\ipykernel_launcher.py:18: RuntimeWarning: invalid value encountered in longlong_scalars\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0 0 0 ... 1 1 1]\n",
      "[1 1 1 ... 1 1 1]\n",
      "[1 1 1 ... 1 1 1]\n",
      "[1 1 1 ... 1 1 1]\n",
      "[0 0 1 ... 1 1 1]\n",
      "[1 1 1 ... 1 1 1]\n",
      "[1 1 1 ... 1 1 1]\n",
      "[1 1 1 ... 1 1 1]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\ipykernel_launcher.py:18: RuntimeWarning: invalid value encountered in longlong_scalars\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\ipykernel_launcher.py:18: RuntimeWarning: invalid value encountered in longlong_scalars\n"
     ]
    }
   ],
   "source": [
    "# Perform anomaly detection and calculate metrics for Test Dataset 1\n",
    "# Initialize dictionaries to store the metrics and prediction times\n",
    "metrics_01 = {'OCSVM': {}, 'PCA': {}, 'kNN': {}, 'Isolation Forest': {}}\n",
    "prediction_times_01 = {'OCSVM': 0, 'PCA': 0, 'kNN': 0, 'Isolation Forest': 0}\n",
    "\n",
    "for model_name, model in [('OCSVM', best_ocsvm), ('PCA', best_pca), ('kNN', best_knn), ('Isolation Forest', best_iforest)]:\n",
    "    start_time = time.time()\n",
    "    y_pred_01 = model.predict(X_test_01)\n",
    "    end_time = time.time()\n",
    "    prediction_time = end_time - start_time\n",
    "\n",
    "    prediction_times_01[model_name] = prediction_time\n",
    "    prediction_times_01[model_name] = round(prediction_time, 4)  # Round to 4 decimal places\n",
    "\n",
    "\n",
    "    y_true_01 = np.ones_like(y_pred_01)\n",
    "    tn, fp, fn, tp = confusion_matrix(y_true_01, y_pred_01, labels=[0, 1]).ravel()\n",
    "    fpr = fp / (fp + tn)\n",
    "    \n",
    "    print(y_pred_01)\n",
    "    print(y_true_01)\n",
    "    metrics_01[model_name]['Accuracy'] = accuracy_score(y_true_01, y_pred_01)\n",
    "    metrics_01[model_name]['Precision'] = precision_score(y_true_01, y_pred_01)\n",
    "    metrics_01[model_name]['Recall'] = recall_score(y_true_01, y_pred_01)\n",
    "    metrics_01[model_name]['F1 Score'] = f1_score(y_true_01, y_pred_01)\n",
    "    metrics_01[model_name]['False Positive Rate'] = fpr\n",
    "\n",
    "    metrics_01[model_name]['Accuracy'] = round(metrics_01[model_name]['Accuracy'], 6)  # Round to 6 decimal places\n",
    "    metrics_01[model_name]['Precision'] = round(metrics_01[model_name]['Precision'], 6)  # Round to 6 decimal places\n",
    "    metrics_01[model_name]['Recall'] = round(metrics_01[model_name]['Recall'], 6)  # Round to 6 decimal places\n",
    "    metrics_01[model_name]['F1 Score'] = round(metrics_01[model_name]['F1 Score'], 6)  # Round to 6 decimal places\n",
    "    metrics_01[model_name]['False Positive Rate'] = round(fpr, 6)  # Round to 6 decimal places\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6f2bf5d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\ipykernel_launcher.py:19: RuntimeWarning: invalid value encountered in longlong_scalars\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\ipykernel_launcher.py:19: RuntimeWarning: invalid value encountered in longlong_scalars\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\ipykernel_launcher.py:19: RuntimeWarning: invalid value encountered in longlong_scalars\n",
      "C:\\ProgramData\\anaconda3\\envs\\py36\\lib\\site-packages\\ipykernel_launcher.py:19: RuntimeWarning: invalid value encountered in longlong_scalars\n"
     ]
    }
   ],
   "source": [
    "# Perform anomaly detection and calculate metrics for Test Dataset 2\n",
    "# Initialize dictionaries to store the metrics\n",
    "metrics_02 = {'OCSVM': {}, 'PCA': {}, 'kNN': {}, 'Isolation Forest': {}}\n",
    "prediction_times_02 = {'OCSVM': 0, 'PCA': 0, 'kNN': 0, 'Isolation Forest': 0}\n",
    "\n",
    "# Calculate metrics for each model\n",
    "for model_name, model in [('OCSVM', best_ocsvm), ('PCA', best_pca), ('kNN', best_knn), ('Isolation Forest', best_iforest)]:\n",
    "    start_time = time.time()\n",
    "    y_pred_02 = model.predict(X_test_02)\n",
    "    end_time = time.time()\n",
    "    prediction_time = end_time - start_time\n",
    "\n",
    "    prediction_times_02[model_name] = prediction_time\n",
    "    prediction_times_02[model_name] = round(prediction_time, 4)  # Round to 4 decimal places\n",
    "\n",
    "\n",
    "    y_true_02 = np.ones_like(y_pred_02)\n",
    "    tn, fp, fn, tp = confusion_matrix(y_true_02, y_pred_02, labels=[0, 1]).ravel()\n",
    "    fpr = fp / (fp + tn)\n",
    "    \n",
    "    metrics_02[model_name]['Accuracy'] = accuracy_score(y_true_02, y_pred_02)\n",
    "    metrics_02[model_name]['Precision'] = precision_score(y_true_02, y_pred_02)\n",
    "    metrics_02[model_name]['Recall'] = recall_score(y_true_02, y_pred_02)\n",
    "    metrics_02[model_name]['F1 Score'] = f1_score(y_true_02, y_pred_02)\n",
    "    metrics_02[model_name]['False Positive Rate'] = fpr\n",
    "    metrics_02[model_name]['Accuracy'] = round(metrics_02[model_name]['Accuracy'], 6)  # Round to 6 decimal places\n",
    "    metrics_02[model_name]['Precision'] = round(metrics_02[model_name]['Precision'], 6)  # Round to 6 decimal places\n",
    "    metrics_02[model_name]['Recall'] = round(metrics_02[model_name]['Recall'], 6)  # Round to 6 decimal places\n",
    "    metrics_02[model_name]['F1 Score'] = round(metrics_02[model_name]['F1 Score'], 6)  # Round to 6 decimal places\n",
    "    metrics_02[model_name]['False Positive Rate'] = round(fpr, 6)  # Round to 6 decimal places\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "60341f15",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results for Test Dataset 1:\n",
      "OCSVM\n",
      "Accuracy: 0.99393\n",
      "Precision: 1.0\n",
      "Recall: 0.99393\n",
      "F1 Score: 0.996956\n",
      "False Positive Rate: nan\n",
      "Prediction Time: 0.047 seconds\n",
      "\n",
      "PCA\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "F1 Score: 1.0\n",
      "False Positive Rate: nan\n",
      "Prediction Time: 0.0 seconds\n",
      "\n",
      "kNN\n",
      "Accuracy: 0.998403\n",
      "Precision: 1.0\n",
      "Recall: 0.998403\n",
      "F1 Score: 0.999201\n",
      "False Positive Rate: nan\n",
      "Prediction Time: 0.2339 seconds\n",
      "\n",
      "Isolation Forest\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "F1 Score: 1.0\n",
      "False Positive Rate: nan\n",
      "Prediction Time: 0.0625 seconds\n",
      "\n",
      "Results for Test Dataset 2:\n",
      "OCSVM\n",
      "Accuracy: 0.906897\n",
      "Precision: 1.0\n",
      "Recall: 0.906897\n",
      "F1 Score: 0.951175\n",
      "False Positive Rate: nan\n",
      "Prediction Time: 0.002 seconds\n",
      "\n",
      "PCA\n",
      "Accuracy: 1.0\n",
      "Precision: 1.0\n",
      "Recall: 1.0\n",
      "F1 Score: 1.0\n",
      "False Positive Rate: nan\n",
      "Prediction Time: 0.001 seconds\n",
      "\n",
      "kNN\n",
      "Accuracy: 0.991379\n",
      "Precision: 1.0\n",
      "Recall: 0.991379\n",
      "F1 Score: 0.995671\n",
      "False Positive Rate: nan\n",
      "Prediction Time: 0.0453 seconds\n",
      "\n",
      "Isolation Forest\n",
      "Accuracy: 0.824138\n",
      "Precision: 1.0\n",
      "Recall: 0.824138\n",
      "F1 Score: 0.903592\n",
      "False Positive Rate: nan\n",
      "Prediction Time: 0.0277 seconds\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Print the results\n",
    "print(\"Results for Test Dataset 1:\")\n",
    "for model_name, metrics in metrics_01.items():\n",
    "    print(model_name)\n",
    "    for metric_name, value in metrics.items():\n",
    "        print(metric_name + \":\", value)\n",
    "    print(\"Prediction Time:\", prediction_times_01[model_name],\"seconds\")\n",
    "    print()\n",
    "\n",
    "print(\"Results for Test Dataset 2:\")\n",
    "for model_name, metrics in metrics_02.items():\n",
    "    print(model_name)\n",
    "    for metric_name, value in metrics.items():\n",
    "        print(metric_name + \":\", value)\n",
    "    print(\"Prediction Time:\", prediction_times_02[model_name],\"seconds\")\n",
    "    print()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18cf5b0a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
